dss/storage/bundles.py

```diff
@@ -0,0 +1,65 @@
+import re
+import typing
+
+BUNDLE_PREFIX = "bundles"
+FILE_PREFIX = "files"
+TOMBSTONE_SUFFIX = "dead"
+
+UUID_PATTERN = "[0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-4[0-9A-Fa-f]{3}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{12}"
+UUID_REGEX = re.compile(UUID_PATTERN)
+
+VERSION_PATTERN = "\d{4}-\d{2}-\d{2}T\d{2}\d{2}\d{2}[.]\d{6}Z"
+VERSION_REGEX = re.compile(VERSION_PATTERN)
+
+# matches just fully qualified bundle identifiers
+DSS_BUNDLE_FQID_PATTERN = f"({UUID_PATTERN})\.({VERSION_PATTERN})"
+DSS_BUNDLE_FQID_REGEX = re.compile(DSS_BUNDLE_FQID_PATTERN)
+
+# matches just bundle keys
+DSS_BUNDLE_KEY_REGEX = re.compile(f"^{BUNDLE_PREFIX}/{DSS_BUNDLE_FQID_PATTERN}$")
+# matches just bundle tombstones
+DSS_BUNDLE_TOMBSTONE_REGEX = re.compile(
+    f"^{BUNDLE_PREFIX}/({UUID_PATTERN})(?:\.(" + VERSION_PATTERN + "))?\." + TOMBSTONE_SUFFIX + "$")
+# matches all bundle objects
+DSS_OBJECT_NAME_REGEX = re.compile(
+    f"^{BUNDLE_PREFIX}/({UUID_PATTERN})(?:\.({VERSION_PATTERN}))?(?:\.{TOMBSTONE_SUFFIX})?$")
+
+
+def bundle_key_to_bundle_fqid(bundle_key: str) -> str:
```

Comment:
I would like to propose introducing a class to represent bundle IDs:

```python
class BundleID(namedtuple('BundleID', 'uuid version')):
    @classmethod
    def from_key(cls, bundle_key: str):
        match = DSS_OBJECT_NAME_REGEX.match(bundle_key)
        uuid, version, tombstone_suffix = match.groups() if match else (None, None, None)
        if not tombstone_suffix:
            if uuid and version:
                return cls(uuid, version)
        else:
            if uuid:
                return TombstoneID(uuid=uuid, version=version or None)
        raise ValueError(f"Object name does not contain a valid bundle identifier: {bundle_key}")

    def to_key(self, prefix=BUNDLE_PREFIX):
        return f"{prefix}/{self}"

    def to_file_key(self):
        return self.to_key(prefix=FILE_PREFIX)

    def __str__(self):
        return f"{self.uuid}.{self.version}"

class TombstoneID(BundleID):
    def to_key_prefix(self, prefix=BUNDLE_PREFIX):
        return f"{prefix}/{self.uuid}.{self.version or ''}"

    def __str__(self):
        if self.version:
            return f"{self.uuid}.{self.version}.{TOMBSTONE_SUFFIX}"
        else:
            return f"{self.uuid}.{TOMBSTONE_SUFFIX}"
```

By parsing a key into an immutable object once and from that point forward only dealing with that object, you avoid having to constantly validate arguments when composing the various string forms (key, fqid, etc).

Ultimately, `BundleDocument.bundle_id` should be a value of type `BundleID` or `TombstoneID`.

Reply:
I agree with this. There are already four PRs in flight right now, so, if possible, I'd like to implement it after this is merged in a separate refactor PR. Sounds good?

Reply:
Yup.